### PR TITLE
Added initial implementation

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/runtime/runtime.h [hunk 1]
```diff
@@ -38,13 +38,17 @@ typedef enum {
...
+  RESOURCE_ALLOCATION_STRATEGY_IDX,
   SERIALIZATION_LEN, // NEVER USED FOR DATA, USED TO DETERMINE LENGTH OF SERIALIZED INFO
 } SerializedInfoIndex;
 
 std::string base64_encode(const std::string& in);
 std::string base64_decode(const std::string& in);
 std::string serialize_bindings(const std::vector<std::string>& bindings);
 
+std::string resource_allocation_strategy_to_string(TRTEngine::ResourceAllocationStrategy strategy);
+TRTEngine::ResourceAllocationStrategy resource_allocation_strategy_from_string(const std::string& str);
```
Comment (narendasan) [inline core/runtime/runtime.h]: Forgot to delete these

File: core/runtime/runtime.h [hunk 2]
```diff
@@ -38,13 +38,17 @@ typedef enum {
   SERIALIZED_METADATA_IDX,
   TARGET_PLATFORM_IDX,
   REQUIRES_OUTPUT_ALLOCATOR_IDX,
+  RESOURCE_ALLOCATION_STRATEGY_IDX,
```
Comment (narendasan) [inline core/runtime/runtime.h:41]: Make sure to bump the ABI version

Comment (cehongwang) [reply core/runtime/runtime.h:41]: Do you mean changing ABI_VERSION=7 to 8?

Comment (narendasan) [reply core/runtime/runtime.h:41]: yeah

**Review by narendasan — COMMENTED** [review]
Can you add some tests cases for the new API at least for Python?

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 3]
```diff
@@ -184,6 +188,11 @@ def _pack_engine_info(self) -> List[str | bytes]:
         engine_info[REQUIRES_OUTPUT_ALLOCATOR_IDX] = str(
             int(self.requires_output_allocator)
         )
+        print(f"PROVIDED RESOURCE ALLOCATION STRATEGY: {self.dynamically_allocate_resources}")
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py]: This should be in the logging or removed

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 4]
```diff
@@ -184,6 +188,11 @@ def _pack_engine_info(self) -> List[str | bytes]:
...
+        engine_info[RESOURCE_ALLOCATION_STRATEGY_IDX] = str(
+            int(self.dynamically_allocate_resources)
+        )
+        print(engine_info[RESOURCE_ALLOCATION_STRATEGY_IDX])
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py]: Logging or removed

File: tests/py/dynamo/runtime/test_005_dynamic_allocation.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import unittest
+
+import torch
+import torch_tensorrt as torch_trt
+from torch.testing._internal.common_utils import TestCase
+from torch_tensorrt.dynamo.utils import COSINE_THRESHOLD, cosine_similarity
+
+assertions = unittest.TestCase()
+
+if importlib.util.find_spec("torchvision"):
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_005_dynamic_allocation.py]: If torchvision is not critical dont use it. I think a hand defined model will be just as good here
